services/dynamic-sidecar/src/simcore_service_dynamic_sidecar/core/docker_utils.py

```diff
@@ -205,57 +208,57 @@ def _parse_docker_pull_progress(
-            )
-        elif status == _TargetPullStatus.DOWNLOAD_COMPLETE:
-            # reads
-            _, layer_total_size = image_pulling_data[layer_id]
-            # writes
-            image_pulling_data[layer_id] = (
-                round(_DOWNLOAD_RATIO * layer_total_size),
-                layer_total_size,
-            )
-        elif status == _TargetPullStatus.EXTRACTING:
-            # reads
-            _, layer_total_size = image_pulling_data[layer_id]
-
-            # writes
-            assert "current" in docker_pull_progress["progressDetail"]  # nosec
-            image_pulling_data[layer_id] = (
-                round(
-                    _DOWNLOAD_RATIO * layer_total_size
-                    + (1 - _DOWNLOAD_RATIO)
-                    * docker_pull_progress["progressDetail"]["current"]
-                ),
-                layer_total_size,
-            )
-        elif status == _TargetPullStatus.PULL_COMPLETE:
-            # reads
-            _, layer_total_size = image_pulling_data[layer_id]
-            # writes
-            image_pulling_data[layer_id] = (
-                layer_total_size,
-                layer_total_size,
-            )
-        return True
+    if status not in _ALL_TARGET_PULL_STATUSES:
+        return False  # no pull progress logged
+
+    progress_detail = docker_pull_progress.get("progressDetail", {})
 
-    return False  # no pull progress logged
+    layer_id: LayerId | None = docker_pull_progress.get("id", None)
+    if layer_id is None:
+        return False
+    # inits (read/write order is not guaranteed)
+    image_pulling_data.setdefault(layer_id, (0, 0))
+
+    if status == _TargetPullStatus.DOWNLOADING:
+        if "current" not in progress_detail or "total" not in progress_detail:
+            return False
+
+        # writes
+        image_pulling_data[layer_id] = (
+            round(_DOWNLOAD_RATIO * progress_detail["current"]),
+            progress_detail["total"],
+        )
+    elif status == _TargetPullStatus.DOWNLOAD_COMPLETE:
+        # reads
+        _, layer_total_size = image_pulling_data[layer_id]
+        # writes
+        image_pulling_data[layer_id] = (
+            round(_DOWNLOAD_RATIO * layer_total_size),
+            layer_total_size,
+        )
+    elif status == _TargetPullStatus.EXTRACTING:
+        if "current" not in progress_detail:
+            return False
+
+        # reads
+        _, layer_total_size = image_pulling_data[layer_id]
+        # writes
+        image_pulling_data[layer_id] = (
+            round(
+                _DOWNLOAD_RATIO * layer_total_size
+                + (1 - _DOWNLOAD_RATIO) * progress_detail["current"]
+            ),
+            layer_total_size,
+        )
+    elif status == _TargetPullStatus.PULL_COMPLETE:
+        # reads
+        _, layer_total_size = image_pulling_data[layer_id]
+        # writes
+        image_pulling_data[layer_id] = (
+            layer_total_size,
+            layer_total_size,
+        )
```

Comment:
this code looks a lot like parsing... which is what pydantic is made for. maybe for next refactoring.